_drafts/ubsan_trap.md

```diff
@@ -0,0 +1,219 @@
+best on Linux (though it is available on some other platforms)
+
+UBSan can catch several types of undefined behavior, for example:
+
+- **array bounds checking**
+- **integer / floating point overflows** (note that these are undefined per the
+  C spec, but are usually deterministic for a particular compiler + arch +
+  lib... however, changing optimization level might change how the UB is
+  expressed!)
+- **alignment when dereferencing**; it's possible to trick a compiler when
+  dereferencing type-punned objects and in other scenarios, this traps those
+  operations, though they may also end up in fault handler anyway depending on
+  your architecture
+- **loading an out-of-range enum**
+
+## Basic Usage
+
+To enable UBSan, add this flag during compilation and linking:
+
+`-fsanitze=undefined`
+
+### Examples
+
+Consider the small C program below:
+
+```c
+// ubsan-basic.c
+
+#include <stdio.h>
+#include <stdlib.h>
+
+int main(int argc, char **argv) {
+  (void)argc;
+
+  const char *arr[] = {
+      "0",
+      "1",
+  };
+
+  // sketchy code below- argv may not have a second entry! could easily have an
+  // out-of-bounds access in arr!
+  int foo = atoi(argv[1]);
+  printf("foo is %s\n", arr[foo]);
+
+  return 0;
+}
+```
+
+#### Without UBsan
+
+Building that program normally:
+
+```bash
+❯ gcc ubsan-basic.c -o ubsan-basic
+```
+
+And running with problematic input:
+
+```bash
+❯ ./ubsan-basic 2
+foo is
+```
+
+We get a weird result! we've indexed into the `arr` array at an out-of-bounds
+position (2), but the program happily fetched that memory location and tried to
+print it as a string 😬.
+
+This type of bug is very popular over at https://cve.mitre.org/ 😔
+
+#### WIth UBsan
+
+Building the same program *with* UBSan:
+
+```bash
+❯ gcc -fsanitize=undefined ubsan-basic.c -o ubsan-basic
+```
+
+We can see that the runtime was linked in:
+
+```bash
+❯ ldd ubsan-basic
+        linux-vdso.so.1 (0x00007ffcccdfd000)
+
+        # libusbsan below
+        libubsan.so.1 => /lib/x86_64-linux-gnu/libubsan.so.1 (0x00007f22c0cb5000)
+
+        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f22c0ac9000)
+        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f22c0ac2000)
+        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f22c0aa0000)
+        libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f22c0887000)
+        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f22c086a000)
+        /lib64/ld-linux-x86-64.so.2 (0x00007f22c163f000)
+        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f22c071c000)
+```
+
+If we run that program with problematic input, we get an error, and our program
+crashes instead of doing something unexpected!
+
+![img/ubsan-trap/ubsan-basic.png](/img/ubsan-trap/ubsan-basic.png)
+
+## `libubsan` library
+
+For hosted platforms (eg Linux, BSD's, macOS, Windows), there's usually a
+`libusban` library installed with your toolchain. On my system it's here:
```

Comment:
Typo: `libubsan`

Reply:
:fire: USB :fire:

Reply:
heh i had 2 counts of USB in this article :sweat_smile: bless u sir